common/src/main/java/com/microsoft/identity/common/internal/platform/RawKeyAccessor.java

```diff
@@ -0,0 +1,237 @@
+        } catch (final InvalidKeyException e) {
+            errCode = INVALID_KEY;
+            exception = e;
+        } catch (final InvalidAlgorithmParameterException e) {
+            errCode = INVALID_ALG_PARAMETER;
+            exception = e;
+        }
+        throw new ClientException(errCode, exception.getMessage(), exception);
+    }
+
+    @Override
+    public byte[] sign(@NonNull final byte[] text, @Nullable final IDevicePopManager.SigningAlgorithm alg) throws ClientException {
+        final String errCode;
+        final Exception exception;
+        try {
+            final SecretKeySpec key = new SecretKeySpec(this.key, suite.cipherName());
+            Mac mac = Mac.getInstance(suite.macName());
+            mac.init(key);
+            return mac.doFinal(text);
+        } catch (final NoSuchAlgorithmException e) {
+            errCode = NO_SUCH_ALGORITHM;
+            exception = e;
+        } catch (final InvalidKeyException e) {
+            errCode = INVALID_KEY;
+            exception = e;
+        }
+        throw new ClientException(errCode, exception.getMessage());
+    }
+
+    @Override
+    public boolean verify(@NonNull final byte[] text, @Nullable final IDevicePopManager.SigningAlgorithm alg,
+                          @NonNull final byte[] signature) throws ClientException {
+        return Arrays.equals(signature, sign(text, alg));
+    }
+
+    @Override
+    public byte[] getThumprint() throws ClientException {
+        return new byte[0];
+    }
+
+    private static byte[] deriveKey(byte[] keyDerivationKey, byte[] fixedInput)
+            throws IOException, NoSuchAlgorithmException, InvalidKeyException {
+        byte ctr;
+        byte[] cHMAC;
+        byte[] keyDerivated;
+        byte[] dataInput;
+
+        int len;
+        int numCurrentElements;
+        int numCurrentElementsBytes;
+        int outputSizeBit = 256;
+
+        numCurrentElements = 0;
+        ctr = 1;
+        keyDerivated = new byte[outputSizeBit / 8];
+        final SecretKeySpec keySpec = new SecretKeySpec(keyDerivationKey, "HmacSHA256");
+        final Mac hmacSHA256 = Mac.getInstance("HmacSHA256");
+
+        do {
+            dataInput = updateDataInput(ctr, fixedInput);
+            hmacSHA256.reset();
+            hmacSHA256.init(keySpec);
+            hmacSHA256.update(dataInput);
+            cHMAC = hmacSHA256.doFinal();
+            if (256 >= outputSizeBit) {
+                len = outputSizeBit;
+            } else {
+                len = Math.min(256, outputSizeBit - numCurrentElements);
+            }
+
+            numCurrentElementsBytes = numCurrentElements / 8;
+            System.arraycopy(cHMAC, 0, keyDerivated, numCurrentElementsBytes, 32);
+            numCurrentElements = numCurrentElements + len;
+            ctr++;
+        } while (numCurrentElements < outputSizeBit);
+        return keyDerivated;
+    }
+
+    public byte[] generateDerivedKey(@NonNull final byte[] label, @NonNull final byte[] ctx)
```

Comment:
Maybe consider alternate naming for `generateDerivedKey` and `deriveKey`? The method signatures are very similar, but one is a delegate function of the other - maybe easy to inadvertantly invoke the wrong one. One is internal, so maybe is OK....

Should we delegate to specialized object to the KDF? Seems plausible we might want these capabilities outside of the context of this object + these actions appear stateless

Reply:
Yeah.  I've pulled it out.  Should consider making it part of the suite class.